rust/otap-dataflow/crates/engine/src/pipeline_ctrl.rs

```diff
@@ -35,15 +36,97 @@ struct TimerState {
 /// - Only one timer per node is supported at a time.
 /// - All data structures are optimized for single-threaded async use.
 /// - The timer_states map consolidates all timer information for efficiency and correctness.
+/// - The combination of `timer_map` and `canceled` ensures correctness and avoids spurious timer
+///   events.
+///
+/// A reusable per-node repeating timer set.
+///
+/// Manages scheduling, cancellation, and expiration for recurrent timers keyed by NodeId.
+/// Optimized for single-threaded async use.
+struct TimerSet {
+    timers: BinaryHeap<Reverse<(Instant, usize)>>,
+    /// Maps node ID to timer state (scheduled time, duration, and cancellation status).
+    timer_states: HashMap<usize, TimerState>,
+}
+
+impl TimerSet {
+    fn new() -> Self {
+        Self {
+            timers: BinaryHeap::new(),
+            timer_states: HashMap::new(),
+        }
+    }
+
+    /// Schedule or replace a repeating timer for node_id.
+    fn start(&mut self, node_id: usize, duration: Duration) {
+        let when = Instant::now() + duration;
+        self.timers.push(Reverse((when, node_id)));
+        let _ = self.timer_states.insert(
+            node_id,
+            TimerState {
+                scheduled_time: when,
+                duration,
+                is_canceled: false,
+            },
+        );
+    }
+
+    /// Cancel an existing timer for node_id.
+    fn cancel(&mut self, node_id: usize) {
+        // Mark the timer as canceled.
+        if let Some(timer_state) = self.timer_states.get_mut(&node_id) {
+            timer_state.is_canceled = true;
+        }
+    }
+
+    /// Peek the next expiration instant, if any.
+    fn next_expiry(&self) -> Option<Instant> {
+        self.timers.peek().map(|Reverse((when, _))| *when)
+    }
+
+    /// Fire all due timers at or before `now`, invoking the provided callback per firing node.
+    /// Reschedules recurring timers automatically when still active.
+    fn fire_due<F: FnMut(&usize)>(&mut self, now: Instant, mut on_fire: F) {
+        while let Some(Reverse((when, node_id))) = self.timers.peek().cloned() {
+            if when > now {
+                break;
+            }
+            // Pop the entry and validate it.
+            let _ = self.timers.pop();
+            if let Some(timer_state) = self.timer_states.get_mut(&node_id) {
+                if !timer_state.is_canceled && timer_state.scheduled_time == when {
+                    // Fire callback
+                    on_fire(&node_id);
+
+                    // Schedule next recurrence
+                    let next_when = Instant::now() + timer_state.duration;
```

Comment:
Should we use `now` passed as a method parameter instead of calling `Instant::now()` for each timer?
```suggestion
                    let next_when = now + timer_state.duration;
```

Reply:
I agree